sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala

```diff
@@ -138,6 +154,21 @@ object CharVarcharUtils extends Logging {
     StructType(fields)
   }
 
+  def getRawSchema(schema: StructType, conf: SQLConf): StructType = {
+    val fields = schema.map { field =>
+      getRawType(field.metadata).map { dt =>
+        if (conf.getConf(SQLConf.CHAR_AS_VARCHAR)) {
+          val metadata = new MetadataBuilder().withMetadata(field.metadata)
+            .remove(CHAR_VARCHAR_TYPE_STRING_METADATA_KEY).build()
```

Comment:
Isn't the metadata used for char and varchar to string conversion? If `CHAR_AS_VARCHAR` is enabled, it means that we also need drop the metadata even the raw type is varchar?

Reply:
Here we are getting the raw schema, which is restored from this special metadata, and we don't need this special metadata anymore.

It's usually no harm to still keep this special metadata, but in this case, it's inconsistent with the returned raw schema, and it's safer to remove it. I can do the same in the old `getRawSchema` as well if you think it's clearer.

Reply:
Okay, then I think it is fine.